synapse/storage/types.py

```diff
@@ -12,30 +12,49 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Iterable, Iterator, List, Optional, Tuple
+from typing import Any, Iterator, List, Mapping, Optional, Sequence, Tuple, Union
 
 from typing_extensions import Protocol
 
 """
 Some very basic protocol definitions for the DB-API2 classes specified in PEP-249
 """
 
+_Parameters = Union[Sequence[Any], Mapping[str, Any]]
```

Comment:
It is unclear to me if PEP249 actually means `Sequence` when it refers to `Sequence` or not. The type shed hints for this use `Iterable`, not `Sequence`: https://github.com/python/typeshed/blob/master/stdlib/sqlite3/dbapi2.pyi#L142

Reply:
I think it's safer to do `Sequence`, the difference between `Iterable` and `Sequence` is that `Iterable` could be a generator or a `set`, `Sequence` cannot, and I'd be willing to bet that most db implementations are betting on an indexable iterable, ergo a sequence, to be passed along.

Though if you still think `Iterable` is better, just say so, but that's my stand.

Reply:
> I'd be willing to bet that most db implementations are betting on an indexable iterable, ergo a sequence, to be passed along.

I would be surprised if they need it to be indexable, but you're correct that generators / sets cannot be used here. I don't think it matters too much, if this works as `Sequence` than that's fine with me.